clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala

```diff
@@ -0,0 +1,180 @@
+package io.treeverse.clients
+
+import org.apache.hadoop.conf.Configuration
+
+import scala.util.control.Breaks._
+import org.apache.hadoop.tools.DistCp
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}
+
+import java.net.URI
+
+class GCBackupAndRestore {}
+
+object GCBackupAndRestore {
+  val S3AccessKeyName = "fs.s3a.access.key"
+  val S3SecretKeyName = "fs.s3a.secret.key"
+  val AzureStorageAccountKeyName = "fs.azure.account.key"
+  lazy val spark = SparkSession.builder().appName("GCBackupAndRestore").getOrCreate()
+  import spark.implicits._
+
+  /** Validates that args are the following:
+   *  1. address of parquet that includes the relative paths of files to backup or restore, created by a gc run
+   *     2. src: the namespace to backup or restore objects from/to (i.e. repo storage namespace, or an external location compatibly)
+   *     3. backup/restore destination: the namespace to backup or restore objects from/to (i.e. an external location or repo storage namespace compatibly)
+   *     4. Object storage type: "s3" or "azure"
+   */
+  def validateArgs(args: Array[String]): Unit = {
+    if (args.length != 4) {
+      Console.err.println(
+        "Usage: ... <objects list location> <src namespace> <dst namespace> <storage type>"
+      )
+      System.exit(1)
+    }
+
+    val storageType = args(3)
+    if (
+      !storageType
+        .equals(StorageUtils.StorageTypeS3) && !storageType.equals(StorageUtils.StorageTypeAzure)
+    ) {
+      Console.err.println(
+        "Invalid <storage type>, must be either \"s3\" or \"azure\""
+      )
+      System.exit(1)
+    }
+  }
+
+  def constructAbsoluteObjectPaths(
+      objectsRelativePathsDF: DataFrame,
+      srcNamespace: String,
+      storageType: String
+  ): Dataset[String] = {
+    var storageNSForFS = ApiClient
+      .translateURI(URI.create(srcNamespace), storageType)
+      .normalize()
+      .toString
+    if (!storageNSForFS.endsWith("/")) {
+      storageNSForFS += "/"
+    }
+
+    objectsRelativePathsDF
+      .select("address")
+      .as[String]
+      .flatMap(x => StorageUtils.concatKeysToStorageNamespace(Seq(x), storageNSForFS, storageType))
```

Comment:
Can we just

```suggestion
      .map(x => StorageUtils.concatKeysToStorageNamespace(x, storageNSForFS, storageType))
```

?

Reply:
Unfortunately not... StorageUtils.concatKeysToStorageNamespace returns Seq[String]